kotest-framework/kotest-framework-engine/src/commonMain/kotlin/io/kotest/engine/test/interceptors/TimeoutInterceptor.kt

```diff
@@ -32,19 +46,67 @@ internal class TimeoutInterceptor(
       return when (val timeout = testCase.config.timeout) {
          null -> test(testCase, scope)
          else -> try {
-            withTimeout(timeout) {
+            withAppropriateTimeout(timeout) {
                test(testCase, scope.withCoroutineContext(coroutineContext))
             }
-         } catch (t: Throwable) {
-            logger.log { Pair(testCase.name.testName, "Caught timeout $t") }
-            TestResult.Error(mark.elapsedNow(), TestTimeoutException(timeout, testCase.name.testName))
+         } catch (t: CancellationException) {
+            if (t is KotestTimeoutCancellationException || t is TimeoutCancellationException) {
+               logger.log { Pair(testCase.name.testName, "Caught timeout $t") }
+               TestResult.Error(mark.elapsedNow(), TestTimeoutException(timeout, testCase.name.testName, t))
+            } else {
+               throw t
+            }
          }
       }
    }
 }
 
+// The implementation copied from Turbine:
+// https://github.com/cashapp/turbine/blob/1.1.0/src/commonMain/kotlin/app/cash/turbine/channel.kt#L93
+private suspend fun <T> withAppropriateTimeout(
+   timeout: Duration,
+   block: suspend CoroutineScope.() -> T,
+): T {
+   return if (coroutineContext[TestCoroutineScheduler] != null) {
+      // withTimeout uses virtual time, which will hang.
+      withWallclockTimeout(timeout, block)
+   } else {
+      withTimeout(timeout, block)
+   }
+}
+
+private suspend fun <T> withWallclockTimeout(
+   timeout: Duration,
+   block: suspend CoroutineScope.() -> T,
+): T = coroutineScope {
+   val blockDeferred = async(start = CoroutineStart.UNDISPATCHED, block = block)
+
+   // Run the timeout on a scope separate from the caller. This ensures that the use of the
+   // Default dispatcher does not affect the use of a TestScheduler and its fake time.
+   @OptIn(DelicateCoroutinesApi::class)
+   val timeoutJob = GlobalScope.launch(Dispatchers.Default) { delay(timeout) }
+
+   select {
+      blockDeferred.onAwait { result ->
+         timeoutJob.cancel()
+         result
+      }
+      timeoutJob.onJoin {
+         blockDeferred.cancel()
+         throw KotestTimeoutCancellationException("Timed out waiting for $timeout")
+      }
+   }
+}
+
+private class KotestTimeoutCancellationException(message: String) : CancellationException(message)
```

Comment:
could we add a comment on why we've added our own exception type

Reply:
Added 👌